* 비지도학습 기반
* (결정 카테고리 없이)

In [14]:
import pandas as pd
import glob, os
import warnings 
warnings.filterwarnings('ignore')
pd.set_option('display.max_colwidth', 700)

# 아래는 제 컴퓨터에서 압축 파일을 풀어 놓은 디렉터리이니, 여러분의 디렉터리를 설정해 주세요  
path = r'./OpinosisDataset1.0/topics'
# path로 지정한 디렉터리 밑에 있는 모든 .data 파일들의 파일명을 리스트로 취합
all_files = glob.glob(os.path.join(path, "*.data"))    
filename_list = []
opinion_text = []

# 개별 파일들의 파일명은 filename_list 리스트로 취합, 
# 개별 파일들의 파일 내용은 DataFrame 로딩 후 다시 string으로 변환하여 opinion_text 리스트로 취합 
for file_ in all_files:
    # 개별 파일을 읽어서 DataFrame으로 생성 
    df = pd.read_table(file_,index_col=None, header=0,encoding='latin1')
    
    # 절대경로로 주어진 file 명을 가공. 만일 Linux에서 수행시에는 아래 \\를 / 변경. 
    # 맨 마지막 .data 확장자도 제거
    filename_ = file_.split('/')[-1]
    filename = filename_.split('.')[0]

    # 파일명 리스트와 파일 내용 리스트에 파일명과 파일 내용을 추가. 
    filename_list.append(filename)
    opinion_text.append(df.to_string())

# 파일명 리스트와 파일 내용 리스트를  DataFrame으로 생성
document_df = pd.DataFrame({'filename':filename_list, 'opinion_text':opinion_text})
document_df.head()

filename  \
0  rooms_bestwestern_hotel_sfo   
1               speed_windows7   
2   service_holiday_inn_london   
3          video_ipod_nano_8gb   
4      food_holiday_inn_london   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  opinion_text  
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       Great Location ,  Nice   Rooms ,  H...  
1                                                                                                                        Windows 7 is quite simply faster, more stable, boots faster, goes to sleep faster, comes back from sleep faster, manages your files better and on top of that it's beautiful to look at and easy to use .\n0                                                                                                                                                                                                                                           , faster about 20% to 30% faster at running applications than my Vista ,  seriously\n1                                                     ...  
2                                                                                                                                                                                                                                                                                                      not customer, oriented hotelvery low service levelboor reception\n0                                                                                                                                                                                                                                                                    The room was quiet, clean, the bed and pillows were comfortable, and the serv...  
3                                                                                                                                                                                                                                                                           I bought the 8, gig Ipod Nano that has the built, in video camera .\n0                                                                                                                                                                                                                      Itunes has an on, line store, where you may purchase and download music and videos which will install onto the ipod .\n1                           ...  
4                                                                                                                                                                                                                                                                      The room was packed to capaci

In [15]:
# import nltk
# nltk.download()

In [16]:
# from nltk.corpus import brown
# brown.words()

In [17]:
# !python -m nltk.downloader popularimport nltk; nltk.download('popular')

### TfidfVectorizer의 tokenizer인자로 사용될 lemmatization 어근 변환 함수를 설정.


In [18]:
from nltk.stem import WordNetLemmatizer
import string
import nltk

remove_punct_dict = dict((ord(punct), None) for punct in string.punctuation)
lemmar = WordNetLemmatizer()

# 입력으로 들어온 token단어들에 대해서 lemmatization 어근 변환. 
def LemTokens(tokens):
    return [lemmar.lemmatize(token) for token in tokens]

# TfidfVectorizer 객체 생성 시 tokenizer인자로 해당 함수를 설정하여 lemmatization 적용
# 입력으로 문장을 받아서 stop words 제거-> 소문자 변환 -> 단어 토큰화 -> lemmatization 어근 변환. 
def LemNormalize(text):
    return LemTokens(nltk.word_tokenize(text.lower().translate(remove_punct_dict)))

TF-IDF 기반 Vectorization 적용 및 KMeans 군집화 수행.
* Stemming과 Lemmatization 같은 어근 변환은 TfidfVectorizer에서 직접 지원하진 않으나 tokenizer 파라미터에 커스텀 어근 변환 함수를 적용하여 어근 변환을 수행할 수 있음
* TfidfVectorizer 생성자의 tokenizer인자로 위에서 생성 LemNormalize 함수 설정.

In [21]:
# import nltk
# nltk.download('punkt')
# nltk.download('punkt_tab')
# nltk.download('wordnet')

[nltk_data] Downloading package punkt to /aiffel/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package punkt_tab to /aiffel/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.
[nltk_data] Downloading package wordnet to /aiffel/nltk_data...


True

In [23]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vect = TfidfVectorizer(tokenizer=LemNormalize, stop_words='english' , \
                             ngram_range=(1,2), min_df=0.05, max_df=0.85 )

#opinion_text 컬럼값으로 feature vectorization 수행
feature_vect = tfidf_vect.fit_transform(document_df['opinion_text'])

In [24]:
import pandas as pd
pd.DataFrame(feature_vect)

0
0     (0, 3516)\t0.004412514497639796\n  (0, 240)\t0.006378037666540962\n  (0, 496)\t0.0056526215462335975\n  (0, 4344)\t0.006378037666540962\n  (0, 1539)\t0.0053768311369804265\n  (0, 3346)\t0.006378037666540962\n  (0, 1754)\t0.0053768311369804265\n  (0, 680)\t0.006378037666540962\n  (0, 1275)\t0.006378037666540962\n  (0, 742)\t0.0056526215462335975\n  (0, 3907)\t0.006378037666540962\n  (0, 1880)\t0.006378037666540962\n  (0, 4348)\t0.0053768311369804265\n  (0, 3884)\t0.006378037666540962\n  (0, 1011)\t0.006378037666540962\n  (0, 4396)\t0.006378037666540962\n  (0, 1002)\t0.006378037666540962\n  (0, 2048)\t0.006378037666540962\n  (0, 4449)\t0.004412514497639796\n  (0, 4385)\t0.006656158361468...
1     (0, 772)\t0.014100392904325968\n  (0, 2564)\t0.01321779642619888\n  (0, 2315)\t0.014100392904325968\n  (0, 2176)\t0.014100392904325968\n  (0, 2584)\t0.014100392904325968\n  (0, 4349)\t0.014100392904325968\n  (0, 4273)\t0.02499332519137622\n  (0, 2568)\t0.009145357558647847\n  (0, 1865)\t0.01321779642619888\n  (0, 3322)\t0.011358797833426996\n  (0, 4435)\t0.01321779642619888\n  (0, 651)\t0.014100392904325968\n  (0, 4594)\t0.009145357558647847\n  (0, 379)\t0.014100392904325968\n  (0, 3853)\t0.01321779642619888\n  (0, 2503)\t0.01321779642619888\n  (0, 4122)\t0.014100392904325968\n  (0, 1565)\t0.01321779642619888\n  (0, 378)\t0.02381245774119256\n  (0, 0)\t0.02926520257436742\n  (0, 4566)\...
2     (0, 3673)\t0.009548537115369042\n  (0, 2529)\t0.008049634442245858\n  (0, 4368)\t0.008049634442245858\n  (0, 4305)\t0.009548537115369042\n  (0, 3568)\t0.009548537115369042\n  (0, 1647)\t0.008950858363686528\n  (0, 593)\t0.009548537115369042\n  (0, 2831)\t0.008950858363686528\n  (0, 3668)\t0.009548537115369042\n  (0, 739)\t0.009548537115369042\n  (0, 3205)\t0.009548537115369042\n  (0, 4134)\t0.009548537115369042\n  (0, 1982)\t0.008462519266152034\n  (0, 2311)\t0.008950858363686528\n  (0, 2117)\t0.009548537115369042\n  (0, 97)\t0.009548537115369042\n  (0, 3665)\t0.008462519266152034\n  (0, 2740)\t0.009548537115369042\n  (0, 3168)\t0.009548537115369042\n  (0, 1050)\t0.008950858363686528\n...
3     (0, 1926)\t0.004938072413225963\n  (0, 1232)\t0.00437643300012072\n  (0, 727)\t0.004162907605168956\n  (0, 1398)\t0.0039779434948545465\n  (0, 1908)\t0.004162907605168956\n  (0, 1941)\t0.00437643300012072\n  (0, 3275)\t0.004628979939688389\n  (0, 717)\t0.0039779434948545465\n  (0, 3592)\t0.004938072413225963\n  (0, 3092)\t0.004938072413225963\n  (0, 4364)\t0.004938072413225963\n  (0, 2316)\t0.004938072413225963\n  (0, 3151)\t0.003416304081749304\n  (0, 4383)\t0.004938072413225963\n  (0, 4080)\t0.0039779434948545465\n  (0, 1468)\t0.004162907605168956\n  (0, 1933)\t0.0036688510213169724\n  (0, 3249)\t0.004938072413225963\n  (0, 371)\t0.004938072413225963\n  (0, 3743)\t0.00493807241322596...
4     (0, 3421)\t0.0102089948903768\n  (0, 1228)\t0.021781367252297124\n  (0, 4427)\t0.010890683626148562\n  (0, 802)\t0.0102089948903768\n  (0, 3434)\t0.0102089948903768\n  (0, 1035)\t0.010890683626148562\n  (0, 2200)\t0.0102089948903768\n  (0, 4373)\t0.0204179897807536\n  (0, 3704)\t0.009181094544372022\n  (0, 4433)\t0.010890683626148562\n  (0, 1870)\t0.0078003099965891845\n  (0, 3888)\t0.010890683626148562\n  (0, 2746)\t0.010890683626148562\n  (0, 1950)\t0.00877316500445053\n  (0, 733)\t0.010890683626148562\n  (0, 2310)\t0.018362189088744044\n  (0, 1911)\t0.010890683626148562\n  (0, 1475)\t0.010890683626148562\n  (0, 1914)\t0.009181094544372022\n  (0, 1882)\t0.010890683626148562\n  (0, 19...
5     (0, 4263)\t0.024151954353249856\n  (0, 4450)\t0.022640192943713194\n  (0, 4393)\t0.018658033514028192\n  (0, 1324)\t0.024151954353249856\n  (0, 2686)\t0.024151954353249856\n  (0, 3314)\t0.024151954353249856\n  (0, 782)\t0.024151954353249856\n  (0, 383)\t0.024151954353249856\n  (0, 3305)\t0.024151954353249856\n  (0, 2969)\t0.024151954353249856\n  (0, 1668)\t0.021404993933639024\n  (0, 2345)\t0.021404993933639024\n  (0, 868)\t0.04830390870649971\n  (0, 367)\t0.022640192

In [25]:
from sklearn.cluster import KMeans

# 5개 집합으로 군집화 수행. 예제를 위해 동일한 클러스터링 결과 도출용 random_state=0 
km_cluster = KMeans(n_clusters=5, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_
cluster_centers = km_cluster.cluster_centers_

In [26]:
cluster_label

array([0, 4, 0, 3, 0, 4, 3, 1, 1, 0, 2, 1, 0, 0, 1, 3, 2, 0, 0, 2, 1, 2,
       2, 1, 0, 1, 3, 2, 0, 3, 3, 0, 4, 0, 2, 0, 3, 3, 3, 4, 2, 2, 1, 3,
       0, 0, 4, 1, 4, 2, 0], dtype=int32)

In [27]:
cluster_centers

array([[0.        , 0.00099499, 0.00174637, ..., 0.        , 0.00183397,
        0.00144581],
       [0.01065711, 0.        , 0.        , ..., 0.01520495, 0.        ,
        0.        ],
       [0.        , 0.00092551, 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.00973837, 0.        , 0.        , ..., 0.00397274, 0.        ,
        0.        ],
       [0.00967213, 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [28]:
document_df['cluster_label'] = cluster_label
document_df.head()

filename  \
0  rooms_bestwestern_hotel_sfo   
1               speed_windows7   
2   service_holiday_inn_london   
3          video_ipod_nano_8gb   
4      food_holiday_inn_london   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  opinion_text  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       Great Location ,  Nice   Rooms ,  H...   
1                                                                                                                        Windows 7 is quite simply faster, more stable, boots faster, goes to sleep faster, comes back from sleep faster, manages your files better and on top of that it's beautiful to look at and easy to use .\n0                                                                                                                                                                                                                                           , faster about 20% to 30% faster at running applications than my Vista ,  seriously\n1                                                     ...   
2                                                                                                                                                                                                                                                                                                      not customer, oriented hotelvery low service levelboor reception\n0                                                                                                                                                                                                                                                                    The room was quiet, clean, the bed and pillows were comfortable, and the serv...   
3                                                                                                                                                                                                                                                                           I bought the 8, gig Ipod Nano that has the built, in video camera .\n0                                                                                                                                                                                                                      Itunes has an on, line store, where you may purchase and download music and videos which will install onto the ipod .\n1                           ...   
4                                                                                                                                                                                                                                                                      The room was packed to c

In [29]:
document_df[document_df['cluster_label']==0].sort_values(by='filename')

filename  \
17   bathroom_bestwestern_hotel_sfo   
4           food_holiday_inn_london   
50           food_swissotel_chicago   
13       free_bestwestern_hotel_sfo   
12   location_bestwestern_hotel_sfo   
9       location_holiday_inn_london   
33    parking_bestwestern_hotel_sfo   
31         price_holiday_inn_london   
35          room_holiday_inn_london   
0       rooms_bestwestern_hotel_sfo   
44          rooms_swissotel_chicago   
18    service_bestwestern_hotel_sfo   
2        service_holiday_inn_london   
45  service_swissotel_hotel_chicago   
24      staff_bestwestern_hotel_sfo   
28          staff_swissotel_chicago   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
17                                                                                                                                                                                                                                                The room was not overly big, but clean and very comfortable beds, a great shower and very clean bathrooms .\n0                                                                                                                                                                                                                          The second room was smaller, with a very inconvenient bathroom layout, but at least it was quieter and we were able to sleep .\n1 ...   
4                                                                                                                                                                                                                                                                       The room was packed to capacity with queues at the food buffets .\n0                                                                                                                                                   The over zealous   staff cleared our unfinished drinks while we were collecting cooked food and movement around the room with plates was difficult in the crowded circumstances .\n1                                         ...   
50                                                                                                                                                                                                                                                                                                                                                                   The food for our event was delicious .\n0                                                                                                                                                                                                                                                                                                              ...   
13                                                                                                                                                                The wine reception is a great idea as it is nice to meet other travellers and great having access to the free Internet access in our room .\n0                                                                                      They also have a computer available with free internet which is a nice bonus but I didn't find that out till the day before we left but was still

In [30]:
document_df[document_df['cluster_label']==1].sort_values(by='filename')

filename  \
23    accuracy_garmin_nuvi_255W_gps   
42  directions_garmin_nuvi_255W_gps   
47     display_garmin_nuvi_255W_gps   
25                features_windows7   
8    satellite_garmin_nuvi_255W_gps   
7       screen_garmin_nuvi_255W_gps   
11       speed_garmin_nuvi_255W_gps   
14     updates_garmin_nuvi_255W_gps   
20       voice_garmin_nuvi_255W_gps   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
23                                                                                                                                                                                                                                                                 , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...   
42                                                                                                                                                                                                                                                                You also get upscale features like spoken directions including street names and programmable POIs .\n0                                                                                                                                                                                                                                                                                                I used to hesitate to go out of my directions but no...   
47                                                                                                                                                                                                                                                       3 quot  widescreen display was a bonus .\n0                                                                                                                                           This made for smoother graphics on the 255w of the vehicle moving along displayed roads, where the 750's display was more of a  jerky  movement .\n1                                                                                                                         ...   
25                                                                                                                                                                             I had to uninstall anti, virus and selected other programs, some of which did not have listings in the  Programs and Features  Control Panel section .\n0                                                                                                                                                                                                                           This review briefly touches upon some of the key features and enhancements of Microsoft's latest OS .\n1                                                 ...   
8                                                                                                    

In [31]:
document_df[document_df['cluster_label']==2].sort_values(by='filename')

filename  \
27       comfort_honda_accord_2008   
34       comfort_toyota_camry_2007   
10   gas_mileage_toyota_camry_2007   
41      interior_honda_accord_2008   
19      interior_toyota_camry_2007   
49       mileage_honda_accord_2008   
22   performance_honda_accord_2008   
21       quality_toyota_camry_2007   
40         seats_honda_accord_2008   
16  transmission_toyota_camry_2007   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
27                                                                                                                                                                                                                                                                                                                                                                                                Drivers seat not comfortable, the car itself compared to other models of similar class .\n0                                                                                                                                                                                                                               ...   
34                                                                                                                                                  Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                        Seats are fine, in fact of all the smaller sedans this is the most comfortable I found for the price as I am 6', 2  and 250# .\n1                                                                                                                                                                                     Great gas mileage and comfortable on long trips ...   
10                                                                                                                                                                                                                                                                                                                                                                              Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                                                                                                                                                                  ...   
41                                                                                                                                                                                                                                                                                                                                                                                                                                                                             I love the new body style and the interior is a simple pleasure except for the center dash .\n0                                                                                                                                              ...   
19                                                                      

In [32]:
document_df[document_df['cluster_label']==3].sort_values(by='filename')

filename  \
30   battery-life_amazon_kindle   
43   battery-life_ipod_nano_8gb   
6   battery-life_netbook_1005ha   
38      keyboard_netbook_1005ha   
29   performance_netbook_1005ha   
36         screen_ipod_nano_8gb   
15        screen_netbook_1005ha   
26     size_asus_netbook_1005ha   
37          sound_ipod_nano_8gb   
3           video_ipod_nano_8gb   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
30                                                                                                                                                                                                               After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...   
43                                                                                                                                                                                                                                                                                                 short battery life  I moved up from an 8gb .\n0                                                                                                                                                                                                                                                                                            I love this ipod except for the battery life .\n1                             ...   
6                                                                                                                                                                                                                                                                                                                   6GHz 533FSB cpu, glossy display, 3, Cell 23Wh Li, ion Battery  , and a 1 .\n0                                                                                                                                                                                                                                                                                              Not to mention that as of now...   
38                                                                                                                                                                                  ,  I think the new keyboard rivals the great hp mini keyboards .\n0                                                                                                                                 Since the battery life difference is minimum, the only reason to upgrade would be to get the better keyboard .\n1                                                                                                                                                                                   The keyboard is now as good as t...   
29                                                                                                    

In [33]:
document_df[document_df['cluster_label']==4].sort_values(by='filename')

filename  \
39          buttons_amazon_kindle   
32  eyesight-issues_amazon_kindle   
46            fonts_amazon_kindle   
5        navigation_amazon_kindle   
48            price_amazon_kindle   
1                  speed_windows7   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
39                                                                                                                                             I thought it would be fitting to christen my Kindle with the Stephen King novella UR, so went to the Amazon site on my computer and clicked on the button to buy it .\n0                                                                                                                                                                                                            As soon as I'd clicked the button to confirm my order it appeared on my Kindle almost immediately !\n1                                                                                   ...   
32                                                                                                                                                                                            It feels as easy to read as the K1 but doesn't seem any crisper to my eyes   .\n0                                                                                                                                                            the white is really GREY, and to avoid considerable eye, strain I had to refresh pages   every other page .\n1                                    The dream has always been a portable electronic device that could hold a ton of reading material, automate subscriptions and fa...   
46                                                                                                                                                                                                                                                                                                                     Being able to change the font sizes is awesome !\n0                                                                                                                                                                                                                                                                        For whatever reason, Amazon decided to make the Font on the Home Screen   ...   
5                                                                                                                                                                                                                                                                                                                    In fact, the entire navigation structure has been completely revised ,  I'm still getting used to it but it's a huge step forward .\n0                                                                                                                                                                                                                                                                 ...   
48                                                                                                                                                                                                                                 If 

### 군집 3개로 낮춰 보기

In [34]:
from sklearn.cluster import KMeans

# 3개의 집합으로 군집화 
km_cluster = KMeans(n_clusters=3, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_


# 소속 클러스터를 cluster_label 컬럼으로 할당하고 cluster_label 값으로 정렬
document_df['cluster_label'] = cluster_label
document_df.sort_values(by='cluster_label')

filename  \
0       rooms_bestwestern_hotel_sfo   
45  service_swissotel_hotel_chicago   
44          rooms_swissotel_chicago   
35          room_holiday_inn_london   
33    parking_bestwestern_hotel_sfo   
31         price_holiday_inn_london   
28          staff_swissotel_chicago   
24      staff_bestwestern_hotel_sfo   
18    service_bestwestern_hotel_sfo   
17   bathroom_bestwestern_hotel_sfo   
13       free_bestwestern_hotel_sfo   
12   location_bestwestern_hotel_sfo   
50           food_swissotel_chicago   
4           food_holiday_inn_london   
9       location_holiday_inn_london   
2        service_holiday_inn_london   
5          navigation_amazon_kindle   
3               video_ipod_nano_8gb   
36             screen_ipod_nano_8gb   
37              sound_ipod_nano_8gb   
38          keyboard_netbook_1005ha   
39            buttons_amazon_kindle   
42  directions_garmin_nuvi_255W_gps   
43       battery-life_ipod_nano_8gb   
1                    speed_windows7   
46              fonts_amazon_kindle   
47     display_garmin_nuvi_255W_gps   
48              price_amazon_kindle   
32    eyesight-issues_amazon_kindle   
11       speed_garmin_nuvi_255W_gps   
25                features_windows7   
29       performance_netbook_1005ha   
6       battery-life_netbook_1005ha   
14     updates_garmin_nuvi_255W_gps   
15            screen_netbook_1005ha   
26         size_asus_netbook_1005ha   
7       screen_garmin_nuvi_255W_gps   
23    accuracy_garmin_nuvi_255W_gps   
8    satellite_garmin_nuvi_255W_gps   
30       battery-life_amazon_kindle   
20       voice_garmin_nuvi_255W_gps   
16   transmission_toyota_camry_2007   
19       interior_toyota_camry_2007   
40          seats_honda_accord_2008   
41       interior_honda_accord_2008   
22    performance_honda_accord_2008   
49        mileage_honda_accord_2008   
34        comfort_toyota_camry_2007   
27        comfort_honda_accord_2008   
21        quality_toyota_camry_2007   
10    gas_mileage_toyota_camry_2007   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        Great Location ,  Nice   Rooms ,  H...   
45                                                                                                                                                                                                                                                                                                                                                  Mediocre room and service for a very extravagant price .\n0                                                                                                                                                                                         

In [35]:
document_df[document_df['cluster_label']==0].sort_values(by='filename')

filename  \
17   bathroom_bestwestern_hotel_sfo   
4           food_holiday_inn_london   
50           food_swissotel_chicago   
13       free_bestwestern_hotel_sfo   
12   location_bestwestern_hotel_sfo   
9       location_holiday_inn_london   
33    parking_bestwestern_hotel_sfo   
31         price_holiday_inn_london   
35          room_holiday_inn_london   
0       rooms_bestwestern_hotel_sfo   
44          rooms_swissotel_chicago   
18    service_bestwestern_hotel_sfo   
2        service_holiday_inn_london   
45  service_swissotel_hotel_chicago   
24      staff_bestwestern_hotel_sfo   
28          staff_swissotel_chicago   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
17                                                                                                                                                                                                                                                The room was not overly big, but clean and very comfortable beds, a great shower and very clean bathrooms .\n0                                                                                                                                                                                                                          The second room was smaller, with a very inconvenient bathroom layout, but at least it was quieter and we were able to sleep .\n1 ...   
4                                                                                                                                                                                                                                                                       The room was packed to capacity with queues at the food buffets .\n0                                                                                                                                                   The over zealous   staff cleared our unfinished drinks while we were collecting cooked food and movement around the room with plates was difficult in the crowded circumstances .\n1                                         ...   
50                                                                                                                                                                                                                                                                                                                                                                   The food for our event was delicious .\n0                                                                                                                                                                                                                                                                                                              ...   
13                                                                                                                                                                The wine reception is a great idea as it is nice to meet other travellers and great having access to the free Internet access in our room .\n0                                                                                      They also have a computer available with free internet which is a nice bonus but I didn't find that out till the day before we left but was still

In [36]:
document_df[document_df['cluster_label']==1].sort_values(by='filename')

filename  \
23    accuracy_garmin_nuvi_255W_gps   
30       battery-life_amazon_kindle   
43       battery-life_ipod_nano_8gb   
6       battery-life_netbook_1005ha   
39            buttons_amazon_kindle   
42  directions_garmin_nuvi_255W_gps   
47     display_garmin_nuvi_255W_gps   
32    eyesight-issues_amazon_kindle   
25                features_windows7   
46              fonts_amazon_kindle   
38          keyboard_netbook_1005ha   
5          navigation_amazon_kindle   
29       performance_netbook_1005ha   
48              price_amazon_kindle   
8    satellite_garmin_nuvi_255W_gps   
7       screen_garmin_nuvi_255W_gps   
36             screen_ipod_nano_8gb   
15            screen_netbook_1005ha   
26         size_asus_netbook_1005ha   
37              sound_ipod_nano_8gb   
11       speed_garmin_nuvi_255W_gps   
1                    speed_windows7   
14     updates_garmin_nuvi_255W_gps   
3               video_ipod_nano_8gb   
20       voice_garmin_nuvi_255W_gps   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
23                                                                                                                                                                                                                                                                 , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...   
30                                                                                                                                                                                                               After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...   
43                                                                                                                                                                                                                                                                                                 short battery life  I moved up from an 8gb .\n0                                                                                                                                                                                                                                                                                            I love this ipod except for the battery life .\n1                             ...   
6                                                                                                                                                                                       

In [37]:
document_df[document_df['cluster_label']==2].sort_values(by='filename')

filename  \
27       comfort_honda_accord_2008   
34       comfort_toyota_camry_2007   
10   gas_mileage_toyota_camry_2007   
41      interior_honda_accord_2008   
19      interior_toyota_camry_2007   
49       mileage_honda_accord_2008   
22   performance_honda_accord_2008   
21       quality_toyota_camry_2007   
40         seats_honda_accord_2008   
16  transmission_toyota_camry_2007   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
27                                                                                                                                                                                                                                                                                                                                                                                                Drivers seat not comfortable, the car itself compared to other models of similar class .\n0                                                                                                                                                                                                                               ...   
34                                                                                                                                                  Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                        Seats are fine, in fact of all the smaller sedans this is the most comfortable I found for the price as I am 6', 2  and 250# .\n1                                                                                                                                                                                     Great gas mileage and comfortable on long trips ...   
10                                                                                                                                                                                                                                                                                                                                                                              Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                                                                                                                                                                  ...   
41                                                                                                                                                                                                                                                                                                                                                                                                                                                                             I love the new body style and the interior is a simple pleasure except for the center dash .\n0                                                                                                                                              ...   
19                                                                      

### (위) 3개로 줄이니 군집이 잘 됐음

## 군집(Cluster)별 핵심 단어 추출하기
* 각 군집을 구성하는 핵심단어가 어떤 것인지 확인

In [38]:
cluster_centers = km_cluster.cluster_centers_
print('cluster_centers shape :',cluster_centers.shape)
print(cluster_centers)

cluster_centers shape : (3, 4611)
[[0.         0.00099499 0.00174637 ... 0.         0.00183397 0.00144581]
 [0.01005322 0.         0.         ... 0.00706287 0.         0.        ]
 [0.         0.00092551 0.         ... 0.         0.         0.        ]]


(위) 군집이 3개 / word 피쳐가 4611개

각 행의 배열 값은 각 군집 내의 4610개 피처의 위치가 개별 중심과 얼마나 가까운가를 상대 값으로 나타낸 것.

0~1까지의 값을 가질 수 있으며, 1에 가까울 수록 중심과 가까운 값을 의미 함.

### cluster_centers_ 이용, 군집별 핵심 단어 찾기
* ndarray 의 argsort()[: , ::-1]를 이용하여 cluster_centers 배열 내, 값이 큰 순으로 정렬 된 위치 인덱스를 반환. ( 큰 값으로 정렬한 값을 반환하는 것이 아닌 , 큰 값을 가진 배열 내 위치 인덱스 값을 반환 하는 것임 )
* 이 위치 인덱스 값이 필요 한 이유는 핵심 단어 피처의 이름을 출력하기 위함.
* 아래 함수 로직
  * cluster_centers_배열 내에서 가장 값이 큰 데이터의 위치 인덱스를 추출한 뒤 ,  해당 인덱스를 이용해 핵심 단어 이름과 그때의 상대 위치 값을 추출해 , cluster_details 라는 Dict객체 변수에 기록 하고 반환.

In [39]:
# 군집별 top n 핵심단어, 그 단어의 중심 위치 상대값, 대상 파일명들을 반환함. 
def get_cluster_details(cluster_model, cluster_data, feature_names, clusters_num, top_n_features=10):
    cluster_details = {}
    
    # cluster_centers array 의 값이 큰 순으로 정렬된 index 값을 반환
    # 군집 중심점(centroid)별 할당된 word 피처들의 거리값이 큰 순으로 값을 구하기 위함.  
    centroid_feature_ordered_ind = cluster_model.cluster_centers_.argsort()[:,::-1]
    
    #개별 군집별로 iteration하면서 핵심단어, 그 단어의 중심 위치 상대값, 대상 파일명 입력
    for cluster_num in range(clusters_num):
        # 개별 군집별 정보를 담을 데이터 초기화. 
        cluster_details[cluster_num] = {}
        cluster_details[cluster_num]['cluster'] = cluster_num
        
        # cluster_centers_.argsort()[:,::-1] 로 구한 index 를 이용하여 top n 피처 단어를 구함. 
        top_feature_indexes = centroid_feature_ordered_ind[cluster_num, :top_n_features]
        top_features = [ feature_names[ind] for ind in top_feature_indexes ]
        
        # top_feature_indexes를 이용해 해당 피처 단어의 중심 위치 상댓값 구함 
        top_feature_values = cluster_model.cluster_centers_[cluster_num, top_feature_indexes].tolist()
        
        # cluster_details 딕셔너리 객체에 개별 군집별 핵심 단어와 중심위치 상대값, 그리고 해당 파일명 입력
        cluster_details[cluster_num]['top_features'] = top_features
        cluster_details[cluster_num]['top_features_value'] = top_feature_values
        filenames = cluster_data[cluster_data['cluster_label'] == cluster_num]['filename']
        filenames = filenames.values.tolist()
        cluster_details[cluster_num]['filenames'] = filenames
        
    return cluster_details

In [40]:
# 보기 좋게 보기
def print_cluster_details(cluster_details):
    for cluster_num, cluster_detail in cluster_details.items():
        print('####### Cluster {0}'.format(cluster_num))
        print('Top features:', cluster_detail['top_features'])
        print('Reviews 파일명 :',cluster_detail['filenames'][:7])
        print('==================================================')

In [41]:
feature_names = tfidf_vect.get_feature_names()

cluster_details = get_cluster_details(cluster_model=km_cluster, cluster_data=document_df,\
                                  feature_names=feature_names, clusters_num=3, top_n_features=10 )
print_cluster_details(cluster_details)

####### Cluster 0
Top features: ['room', 'hotel', 'service', 'staff', 'food', 'location', 'bathroom', 'clean', 'price', 'parking']
Reviews 파일명 : ['rooms_bestwestern_hotel_sfo', 'service_holiday_inn_london', 'food_holiday_inn_london', 'location_holiday_inn_london', 'location_bestwestern_hotel_sfo', 'free_bestwestern_hotel_sfo', 'bathroom_bestwestern_hotel_sfo']
####### Cluster 1
Top features: ['screen', 'battery', 'keyboard', 'battery life', 'life', 'kindle', 'direction', 'video', 'size', 'voice']
Reviews 파일명 : ['speed_windows7', 'video_ipod_nano_8gb', 'navigation_amazon_kindle', 'battery-life_netbook_1005ha', 'screen_garmin_nuvi_255W_gps', 'satellite_garmin_nuvi_255W_gps', 'speed_garmin_nuvi_255W_gps']
####### Cluster 2
Top features: ['interior', 'seat', 'mileage', 'comfortable', 'gas', 'gas mileage', 'transmission', 'car', 'performance', 'quality']
Reviews 파일명 : ['gas_mileage_toyota_camry_2007', 'transmission_toyota_camry_2007', 'interior_toyota_camry_2007', 'quality_toyota_camry_2007